### PR TITLE
[WIP] Support Diagram

### PR DESCRIPTION
**Pull request by Jocs**
| Q                | A
| ---------------- | ---
| Bug fix?         | no
| New feature?     | yes
| BC breaks?       | no
| Deprecations?    | no
| New tests added? | not needed
| Fixed tickets    | #531
| License          | MIT

### Description

[Description of the bug or feature]

* [x] Editor mermaid
* [x] Copy and paste mermaid
* [x] import and export mermaid
* [x] Editor flowchart
* [x] Copy and paste flowchart
* [x] import and export flowchart
* [x] Editor js-senquence
* [x] Copy and paste js-senquence
* [x] import and export js-senquence
* [x] Editor vega-lite
* [x] Copy and paste vega-lite
* [x] import and export vega-lite

How to use?

Type \@ to quick insert mermaid diagram.

--

### CODE REVIEW

**Comment by Jocs** [comment]
### Test Samples

[vega-lite samples](https://vega.github.io/vega-lite/examples/)

[mermaid samples](https://mermaidjs.github.io/gantt.html)

[flowchart example](https://github.com/adrai/flowchart.js/blob/master/example/index.html)

[sequence github](https://github.com/bramp/js-sequence-diagrams)

### How to insert diagrams?

1. type \@ to open quick insert menu.
2. select diagram type.
3. Input sample code above.

**Comment by Jocs** [comment]
test markdown.md

```markdown
\```vega-lite
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "description": "Horizontally concatenated charts that show different types of discretizing scales.",
  "data": {
    "values": [
      {"a": "A","b": 28}, {"a": "B","b": 55}, {"a": "C","b": 43},
      {"a": "D","b": 91}, {"a": "E","b": 81}, {"a": "F","b": 53},
      {"a": "G","b": 19}, {"a": "H","b": 87}, {"a": "I","b": 52}
    ]
  },
  "hconcat": [
    {
      "mark": "circle",
      "encoding": {
        "y": {
          "field": "b",
          "type": "nominal",
          "sort": null,
          "axis": {
            "ticks": false,
            "domain": false,
            "title": null
          }
        },
        "size": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "quantize"
          }
        },
        "color": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "quantize",
            "zero": true
          },
          "legend": {
            "title": "Quantize"
          }
        }
      }
    },
    {
      "mark": "circle",
      "encoding": {
        "y": {
          "field": "b",
          "type": "nominal",
          "sort": null,
          "axis": {
            "ticks": false,
            "domain": false,
            "title": null
          }
        },
        "size": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "quantile",
            "range": [80, 160, 240, 320, 400]
          }
        },
        "color": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "quantile",
            "scheme": "magma"
          },
          "legend": {
            "format": "d",
            "title": "Quantile"
          }
        }
      }
    },
    {
      "mark": "circle",
      "encoding": {
        "y": {
          "field": "b",
          "type": "nominal",
          "sort": null,
          "axis": {
            "ticks": false,
            "domain": false,
            "title": null
          }
        },
        "size": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "threshold",
            "domain": [30, 70],
            "range": [80, 200, 320]
          }
        },
        "color": {
          "field": "b",
          "type": "quantitative",
          "scale": {
            "type": "threshold",
            "domain": [30, 70],
            "scheme": "viridis"
          },
          "legend": {
            "title": "Threshold"
          }
        }
      }
    }
  ],
  "resolve": {
    "scale": {
      "color": "independent",
      "size": "independent"
    }
  }
}
\```

\```flowchart
st=>start: Start|past:>http://www.google.com[blank]
e=>end: End:>http://www.google.com
op1=>operation: My Operation|past
op2=>operation: Stuff|current
sub1=>subroutine: My Subroutine|invalid
cond=>condition: Yes
or No?|approved:>http://www.google.com
c2=>condition: Good idea|rejected
io=>inputoutput: catch something...|request
para=>parallel: parallel tasks

st->op1(right)->cond
cond(yes, right)->c2
cond(no)->para
c2(true)->io->e
c2(false)->e

para(path1, bottom)->sub1(left)->op1
para(path2, right)->op2->e

st@>op1({"stroke":"Red"})@>cond({"stroke":"Red","stroke-width":6,"arrow-end":"classic-wide-long"})@>c2({"stroke":"Red"})@>op2({"stroke":"Red"})@>e({"stroke":"Red"})
\```

\```sequence
A->B: Does something
\```

\```mermaid
gantt
    title A Gantt Diagram
    dateFormat  YYYY-MM-DD
    section Section
    A task           :a1, 2014-01-01, 30d
    Another task     :after a1  , 20d
    section Another
    Task in sec      :2014-01-12  , 12d
    another task      : 24d
\```
```

**Comment by Jocs** [comment]
Already known bugs:

Input markdown below:

```markdown
\```vega-lite
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "data": {"url": "https://vega.github.io/editor/data/barley.json"},
  "mark": "bar",
  "encoding": {
    "column": {"field": "year","type": "ordinal"},
    "x": {"field": "yield","type": "quantitative","aggregate": "sum"},
    "y": {"field": "variety","type": "nominal"},
    "color": {"field": "site","type": "nominal"}
  }
}
\```

### hello world wor *nice to meet* nice to meet *wowow{cursor}wowowowowowowowoieowowowow*
```

then edit cursor position, the rendered diagram will blink. help is wanted.

**Comment by fxha** [comment]
@Jocs  Do we need additional system dependencies for this PR? Perhaps a little strange but do we need this library because it's reported as optional error?

```
Package cairo was not found in the pkg-config search path.
```

Travis CI log: `warning Error running install script for optional dependency: "/home/travis/build/marktext/marktext/node_modules/canvas: Command failed.` and `info This module is OPTIONAL, you can safely ignore this error`

**Comment by Jocs** [comment]
>and info This module is OPTIONAL, you can safely ignore this error

I think we don't need this, can you run these examples above if you don't install this dependency.

**Comment by fxha** [comment]
Yes, it worked but it's very confusing because there is no optional error message on my machine:

```
Package cairo was not found in the pkg-config search path.
Perhaps you should add the directory containing `cairo.pc'
to the PKG_CONFIG_PATH environment variable
Package 'cairo', required by 'virtual:world', not found
gyp: Call to './util/has_lib.sh freetype' returned exit status 0 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:345:16)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:198:12)
gyp ERR! System Linux 4.18.16
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd marktext/node_modules/canvas
gyp ERR! node -v v8.12.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok

> canvas-prebuilt@1.6.11 install marktext/node_modules/canvas-prebuilt
> node-pre-gyp install

node-pre-gyp WARN Using request for node-pre-gyp https download
[canvas-prebuilt] Success: "marktext/node_modules/canvas-prebuilt/canvas/build/Release/canvas-prebuilt.node" is installed via remote
```

**Comment by Jocs** [comment]
I also didn't see the optional error message too...

File: src/main/menus/view.js [hunk 1]
```diff
@@ -80,7 +80,7 @@ let viewMenu = {
   }]
 }
 
-if (process.env.NODE_ENV !== 'production' || process.env.NODE_ENV === 'development') {
+if (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'development') {
```
[the remark this answers is not in the document]
Comment (Jocs) [reply src/main/menus/view.js]: yes, just for test, and forget change back to `!==`

File: src/muya/lib/config/index.js [hunk 2]
```diff
@@ -211,4 +215,22 @@ export const EXPORT_DOMPURIFY_CONFIG = {
...
+  theme: 'light',
+  markdown: '',
+  preferLooseListItem: true,
+  autoPairBracket: true,
+  autoPairMarkdownSyntax: true,
+  autoPairQuote: true,
+  bulletListMarker: '-',
+  tabSize: 4,
+  sequenceTheme: 'hand', // hand or simple
+  mermaidTheme: 'forest' // dark or forest
```
[the remark this answers is not in the document]
Comment (Jocs) [reply src/muya/lib/config/index.js]: Yes, I'll add it later.

File: src/muya/lib/config/index.js [hunk 3]
```diff
@@ -211,4 +215,22 @@ export const EXPORT_DOMPURIFY_CONFIG = {
...
+}
+
+// export const DIAGRAM_TEMPLATE = {
+//   'mermaid': `graph LR;\nYou-->|Mark Text|Me;`
+// }
```
[the remark this answers is not in the document]
Comment (Jocs) [reply src/muya/lib/config/index.js]: @fxha I originally thought of providing some template text for each Diagram. Later I felt a little redundant, or write it in the document. Do you think it is necessary?

**Comment by Jocs** [comment]
@fxha I found that the crash bug may be caused by this PR, because the previous commit can be run, and the commit cannot be run, but the strange thing is, when I developed this PR, there was no crash?

**Pull request merged**